pkg/volume/flexvolume/driver-call.go

```diff
@@ -0,0 +1,230 @@
+	}
+}
+
+func (dc *DriverCall) Append(arg string) {
+	dc.args = append(dc.args, arg)
+}
+
+func (dc *DriverCall) AppendSpec(spec *volume.Spec, host volume.VolumeHost, extraOptions map[string]string) error {
+	optionsForDriver, err := NewOptionsForDriver(spec, host, extraOptions)
+	if err != nil {
+		return err
+	}
+
+	jsonBytes, err := json.Marshal(optionsForDriver)
+	if err != nil {
+		return fmt.Errorf("Failed to marshal spec, error: %s", err.Error())
+	}
+
+	dc.Append(string(jsonBytes))
+	return nil
+}
+
+func (dc *DriverCall) Run() (*DriverStatus, error) {
+	if dc.plugin.isUnsupported(dc.Command) {
+		return nil, errors.New(StatusNotSupported)
+	}
+	execPath := dc.plugin.getExecutable()
+
+	cmd := dc.plugin.runner.Command(execPath, dc.args...)
+
+	timeout := false
+	if dc.Timeout > 0 {
+		timer := time.AfterFunc(dc.Timeout, func() {
+			timeout = true
+			cmd.Stop()
+		})
+		defer timer.Stop()
+	}
+
+	output, execErr := cmd.CombinedOutput()
+	if execErr != nil {
+		if timeout {
+			return nil, TimeoutError
+		}
+		glog.Errorf("FlexVolume: driver call failed: executable: %s, args: %s, error: %s, output: %s", execPath, dc.args, execErr.Error(), output)
+		_, err := handleCmdResponse(dc.Command, output)
+		if err == nil {
+			glog.Errorf("FlexVolume: driver bug: %s: exec error (%s) but no error in response.", execPath, execErr)
+			return nil, execErr
+		}
+		if isCmdNotSupportedErr(err) {
+			dc.plugin.unsupported(dc.Command)
+		}
+		return nil, err
+	}
+
+	status, err := handleCmdResponse(dc.Command, output)
+	if err != nil {
+		if isCmdNotSupportedErr(err) {
+			dc.plugin.unsupported(dc.Command)
+		}
+		return nil, err
+	}
+
+	return status, nil
+}
+
+// OptionsForDriver represents the spec given to the driver.
+type OptionsForDriver map[string]string
+
+func NewOptionsForDriver(spec *volume.Spec, host volume.VolumeHost, extraOptions map[string]string) (OptionsForDriver, error) {
+	volSource, readOnly := getVolumeSource(spec)
+	options := map[string]string{}
+
+	options[optionFSType] = volSource.FSType
+
+	if readOnly {
+		options[optionReadWrite] = "ro"
+	} else {
+		options[optionReadWrite] = "rw"
+	}
+
+	if volSource.SecretRef != nil {
+		// Extract secret and pass it as options.
+		secrets, err := getSecrets(spec, host)
+		if err != nil {
+			return nil, err
+		}
+		for name, secret := range secrets {
+			options[optionKeySecret+"/"+name] = secret
+		}
+	}
+
+	for key, value := range extraOptions {
+		options[key] = value
+	}
+
+	for key, value := range volSource.Options {
+		options[key] = value
+	}
+
+	return OptionsForDriver(options), nil
+}
+
+// DriverStatus represents the return value of the driver callout.
+type DriverStatus struct {
+	// Status of the callout. One of "Success", "Failure" or "Not supported".
+	Status string
+	// Message is the reason for failure.
+	Message string
+	// Device assigned by the driver.
+	Device string `json:"device"`
+	// The path asked by the plugin
+	Path string `json:"path"`
```

Comment:
Could you make it more clear what the difference between `device` and `path` are (maybe put an example of each). In the previous version `device` was defined as `<Path to the device attached. This field is valid only for attach calls>` so is this a breaking change?